service/stream/exec_data_streamer.go

```diff
@@ -0,0 +1,168 @@
+package stream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"cloud.google.com/go/storage"
+	"github.com/hashicorp/go-multierror"
+	"github.com/onflow/flow-archive/models/archive"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/engine/execution/computation/computer/uploader"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+	"github.com/onflow/flow/protobuf/go/flow/access"
+	execData "github.com/onflow/flow/protobuf/go/flow/executiondata"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type ExecDataStreamer struct {
+	log       zerolog.Logger
+	execApi   execData.ExecutionDataAPIClient
+	accessApi access.AccessAPIClient
+	queue     *archive.SafeDeque // queue of block identifiers for next downloads
+	buffer    *archive.SafeDeque // queue of downloaded execution data records
+	limit     uint               // buffer size limit for downloaded records
+	busy      uint32             // used as a guard to avoid concurrent polling
+	ctx       context.Context
+	chain     flow.ChainID
+}
+
+func NewExecDataStreamer(log zerolog.Logger, accessAddr string, options ...Option) *ExecDataStreamer {
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	ctx := context.Background()
+
+	// initialize clients
+	opts := []grpc.DialOption{grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+		grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(accessAddr, opts...)
+	if err != nil {
+		log.Error()
+	}
+	execDataAPI := execData.NewExecutionDataAPIClient(conn)
+	accessAPI := access.NewAccessAPIClient(conn)
+
+	// get chainID from network params
+	params, err := accessAPI.GetNetworkParameters(ctx, &access.GetNetworkParametersRequest{})
+	if err != nil {
+
+	}
+	chain := flow.ChainID(params.ChainId)
+
+	return &ExecDataStreamer{
+		log:       log,
+		execApi:   execDataAPI,
+		accessApi: accessAPI,
+		queue:     archive.NewDeque(),
+		buffer:    archive.NewDeque(),
+		limit:     cfg.BufferSize,
+		busy:      0,
+		ctx:       ctx,
+		chain:     chain,
+	}
+}
+
+// OnBlockFinalized is a callback for the Flow consensus follower. It is called
+// each time a block is finalized by the Flow consensus algorithm.
+func (e *ExecDataStreamer) OnBlockFinalized(block *model.Block) {
+	blockID := block.BlockID
+	// We push the block ID to the front of the queue; the streamer will try to
+	// download the blocks in a FIFO manner.
+	e.queue.PushFront(blockID)
+
+	e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for download")
+}
+
+// Next returns the next available block data. It returns an ErrUnavailable if no block
+// data is available at the moment.
+func (e *ExecDataStreamer) Next() (*uploader.BlockData, error) {
+	// same implementation as GCPStreamer
+	go e.poll()
```

Comment:
You could simplify this logic a bit by running a fixed size worker pool that continuously reads from `e.queue` fills `e.buffer` up to the configured size. Then there is no need for explicit polling or gating the `poll` method.